src/library/pagination/pagination.component.js

```diff
@@ -0,0 +1,170 @@
+// Import Template
+import template from './pagination.html';
+
+class controller {
+  constructor($log, $scope) {
+    Object.assign(this, {
+      $log,
+      $scope
+    });
+
+    this.currentPage = 0; // binding
+    this.disabled = false; // binding
+    this.itemsPerPage = 0; // binding
+    this.size = ''; // binding
+    this.totalItems = 0; // binding
+    this.visiblePageBuffer = 0; // binding
+
+    this.pages = [];
+    this.totalPages = [];
+  }
+
+  $onInit() {
+    // provide defaults
+    this.currentPage = this.currentPage || 1;
+    this.visiblePageBuffer = this.visiblePageBuffer || this.visiblePageBuffer === 0 ? this.visiblePageBuffer : 3;
+    this.size = this.size || '';
+
+    // validate bindings
+    if (typeof this.itemsPerPage !== 'number' || isFinite(this.itemsPerPage) === false) {
+      this.$log.error('invalid ngbsPagination::itemsPerPage: ', this.itemsPerPage, 'expecting a number');
+    }
+
+    if (this.size && (typeof this.size !== 'string' || ['', 'lg', 'sm'].includes(this.size) === false)) {
+      this.$log.error('invalid ngbsPagination::size: ', this.itemsPerPage, 'expecting sm or lg');
+    }
+
+    if (typeof this.totalItems !== 'number' || isFinite(this.totalItems) === false) {
+      this.$log.error('invalid ngbsPagination::totalItems: ', this.totalItems, 'expecting a number');
+    }
+
+    this.$scope.$watch(() => `${this.currentPage}${this.itemsPerPage}${this.totalItems}${this.visiblePageBuffer}`, () => {
+      this.buildPages();
+    });
+  }
+
+  buildPages() {
+    // determine the number of pages
+    this.totalPages = Math.ceil(this.totalItems / this.itemsPerPage);
```

Comment:
Ah, definitely a terminology difference.  I am totally open to any changes, but I tell you how I got here.

I had 10k photos to display on a screen.  That's too many for 1 screen, so I added pagination.  The original method signature to display the pagination (the series of numbers, representing groups of photos) looked like this (it's PHP, 🤢):

`print_page_index($p_total_pages, $p_current_page_number, $p_width = 4)`

This was great because (IMO) because it was so plainly obvious what it was looking.  If I have 100 photos... and I want to show 10 per screen.... I'd calculate that there should be 10 total pages.  Unfortunately because of the signature, I had to do that calculation every time I wanted to call `print_page_index` (for photos, forum posts, users, etc):

    ceil($total_number_of_photos_available / $total_number_of_photos_to_show_per_screen);

Years later I started using UI Bootstrap and have used it a ton.  Their attributes are identical to what I created here (with the exception of visible-page-buffer (because I took a different approach) and disabled (because I don't like to use `ng-` namespace for components I write).  I loved the fact that I didn't have to run that `ceil()` calculation anymore.  I loved that their *component signature* took arguments that reduced the amount of code I had to write.

I didn't intend to have **item** and **page** used interchangeably (perhaps I wrote some bad documentation).  A page to me (and to UI Bootstrap) is a group of items to be displayed together.  Much like pages of a book are simply groups of words -- where pagination is used to quickly reference a set of words that might be useful (or in the component's case  a set of *items*).  I don't really like the word "items", but I understand its generic to cover anything that the user might want to paginate.

Let me know if that doesn't clear up the attribute names or if you want to discuss further in a better medium (skype/hangouts/slack/discord/email/etc).